### DISCUSSION TITLE
Checkout LineItems Order

### DISCUSSION THREAD

**Discussion by explorador** [Shopify/storefront-api-feedback#32] [Help]
This issue was reported back in 2018.

If user adds Product A, then Product B & then Product C, they would expect to see them in Checkout in that order. However, they see something like: Product C, Product A, Product B which makes them lose track of what product is what.

Issue was originally submitted in the `js-buy-sdk` repository:
https://github.com/Shopify/js-buy-sdk/issues/501

**Comment by mutewinter** [comment]
A workaround I'm using for this right now is to add a `_createdAt` attribute to each line item with the unix epoch time. Then sorting by that on the client side.

_Edit to rename attribute to `_createdAt` to ensure it's hidden during checkout_

**Reply by gabrielvincent** [reply]
Oh, if only it was that simple. This will work fine if you don't care about line item attributes or line items quantities. If you add the exact same item to the cart twice, you'll have two separate items in the cart because they will have different `attributes`.

Your workaround is just the first step of what I had to do to preserve line items order in the cart. Here's a summary of what I had to implement client-side because of this bug in the Shopify Storefront API:

As you suggested, add a `_createdAt` attribute to every item that is added to the cart. The next time a product is added, check the items that are already in the cart to see if any of them match the unique set of `merchandiseId` + `attributes`.
But hey, there will never be a match, since every new item has a different `_createdAt` attribute. Ok, so we need to check the existence of a line item that matches the unique set of `merchandiseId` + `attributes` from the item you are trying to add **ignoring `_createdAt`**.

Here's all the code I had to write to workaround something that should clearly be handle by Shopify server-side:


```typescript
function omitLineItemAttributes(
  lineItem: LineItem,
  attributeKeys: string[]
): LineItem {
  const attributes = (lineItem.attributes ?? []).filter(
    attribute => !attributeKeys.includes(attribute.key)
  )
  return { ...lineItem, attributes }
}

function getLineItem(
  cart: Cart,
  item: LineItem | NewLineItem,
  options?: { ignoreAttributes?: string[] }
) {
  const ignoredAttributes = options?.ignoreAttributes ?? []
  const itemWithFilteredAttributes = omitLineItemAttributes(
    item,
    ignoredAttributes
  )
  return cart.lineItems.find(lineItem => {
    const filteredLineItem = omitLineItemAttributes(lineItem, ignoredAttributes)
    return isSameItems(filteredLineItem, itemWithFilteredAttributes)
  })
}

function isSameItems(itemA?: LineItem, itemB?: LineItem) {
  if (!itemA || !itemB) return false

  return (
    itemA.merchandiseId === itemB.merchandiseId &&
    isEqual(itemA.attributes, itemB.attributes)
  )
}

function getLineItemAttribute(lineItem: LineItem, attributeKey: string) {
  return lineItem.attributes?.find(attribute => attribute.key === attributeKey)
}

function addToCart(
  variant: ProductVariant,
  quantity: number,
  attributes?: Attribute[]
) {
  const protoLineItem: LineItem = {
    merchandiseId: variant.id,
    quantity,
    attributes,
  }
  const currentCart = getCart()
  const existingLineItem = currentCart
    ? getLineItem(currentCart, protoLineItem, {
        ignoreAttributes: ['_createdAt'],
      })
    : null
  const existingAddedAtValue = existingLineItem
    ? getLineItemAttribute(existingLineItem, '_createdAt')?.value
    : null
  const lineItem = {
    ...protoLineItem,
    attributes: [
      ...(protoLineItem.attributes ?? []),
      existingAddedAtValue
        ? { key: '_createdAt', value: existingAddedAtValue }
        : { key: '_createdAt', value: new Date().toISOString() },
    ],
  }

  return graphqlQuery(cartLinesAddMutation, {
    cartId: currentCart.id,
    lines: [lineItem],
  })
}
```
